### PR TITLE
HTTP/2 Prevent modification of activeStreams while iterating

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The Http2Connection interface exposes an activeStreams() method which allows direct iteration over the underlying collection. There are a few places that make copies of this collection to avoid modification while iterating, and a few places that do not make copies. The copy operation can be expensive on hot code paths and also we are not consistently iterating over the activeStreams collection.

Modifications:
- The Http2Connection interface should reduce the exposure of the underlying collection and just expose what is necessary for the interface to function.  This is just a means to iterate over the collection.
- The DefaultHttp2Connection should use this new interface and protect it's internal state while iteration is occurring.

Result:
Reduction in surface area of the Http2Connection interface.  Consistent iteration of the set of active streams.  Concurrent modification exceptions are handled in 1 encapsulated spot.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@nmittler @buchgr - FYI.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 1]
```diff
@@ -142,8 +144,33 @@ public int numActiveStreams() {
...
+            return resultStream;
+        } finally {
+            --activeStreamsIterationCount;
+            if (canModifyActiveStreams()) {
+                for (;;) {
+                    ActiveStreamsEvent event = activeStreamsPendingEvents.poll();
+                    if (event == null) {
+                        break;
+                    }
+                    event.process();
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: This will be part of the "connection closes" if exception happens as part of https://github.com/netty/netty/issues/3501#issuecomment-88213605.

**Comment by Scottmitch** [comment]
Fixes https://github.com/netty/netty/issues/3550.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 2]
```diff
@@ -376,14 +403,10 @@ public Http2Stream open(boolean halfClosed) throws Http2Exception {
...
 
-            if (activeStreams.add(this)) {
-                // Update the number of active streams initiated by the endpoint.
-                createdBy().numActiveStreams++;
-
-                // Notify the listeners.
-                for (int i = 0; i < listeners.size(); i++) {
-                    listeners.get(i).onStreamActive(this);
-                }
+            if (canModifyActiveStreams()) {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Should this all just be done in `addToActiveStreams`?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 3]
```diff
@@ -396,20 +419,13 @@ public Http2Stream close() {
...
 
-                    // Notify the listeners.
-                    for (int i = 0; i < listeners.size(); i++) {
-                        listeners.get(i).onStreamClosed(this);
-                    }
-                } finally {
-                    // Mark this stream for removal.
-                    removalPolicy.markForRemoval(this);
-                }
+            if (canModifyActiveStreams()) {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Should this all be done in `removeFromActiveStreams`?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 4]
```diff
@@ -79,25 +80,15 @@ public void onStreamActive(Http2Stream stream) {
 
     @Override
     public void initialWindowSize(int newWindowSize) throws Http2Exception {
-        int delta = newWindowSize - initialWindowSize;
+        final int delta = newWindowSize - initialWindowSize;
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: why final?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 5]
```diff
@@ -79,25 +80,15 @@ public void onStreamActive(Http2Stream stream) {
...
-                }
-                compositeException.add(e);
-            }
-        }
-        if (compositeException != null) {
-            throw compositeException;
+        LocalActiveStreamVisitor visitor = new LocalActiveStreamVisitor(delta);
+        connection.forEachActiveStream(visitor);
+
+        CompositeStreamException exception = visitor.compositeException();
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: should we just give `visitor` a `throwIfError` (or something similar) method?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 6]
```diff
@@ -391,4 +382,36 @@ void writeWindowUpdate(ChannelHandlerContext ctx) throws Http2Exception {
             ctx.flush();
         }
     }
+
+    /**
+     * Provides a means to iterator over all active streams and increment the flow control windows.
+     */
+    private static final class LocalActiveStreamVisitor implements ActiveStreamVisitor {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: Maybe 'WindowUpdateVisitor'?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java [hunk 7]
```diff
@@ -261,10 +259,25 @@
...
 
     /**
-     * Gets all streams that are actively in use (i.e. {@code OPEN} or {@code HALF CLOSED}). The returned collection is
-     * sorted by priority.
+     * Provide a means of iterating over the collection of active streams.
+     *
+     * @param visitor The visitor which will visit each active stream.
+     * @return The stream before iteration stopped or {@code null} if iteration went past the end.
+     */
+    Http2Stream forEachActiveStream(ActiveStreamVisitor visitor) throws Http2Exception;
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java]: +1

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java [hunk 8]
```diff
@@ -261,10 +259,25 @@
...
      */
-    Collection<Http2Stream> activeStreams();
+    interface ActiveStreamVisitor {
+        /**
+         * @return <ul>
+         *         <li>{@code true} if the processor wants to continue the loop and handle the entry.</li>
+         *         <li>{@code false} if the processor wants to stop handling headers and abort the loop.</li>
+         *         </ul>
+         */
+        boolean visit(Http2Stream stream) throws Http2Exception;
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java]: Similar discussion that we've had prior ... not sure if we need to throw `Http2Exception`.  Really anything that comes back should be some sort of unexpected `RuntimeException`.  We really need to handle `RuntimeException`s anyway.

**Comment by nmittler** [comment]
@Scottmitch looks fantastic, great job!  Just a couple minor nits.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 2 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Humm. We could do this, but then every queued event will re-evalute the conditional via `process()`.  We check the conditional before acting on the queued events so we only have to check 1 time.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 3 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: See https://github.com/netty/netty/pull/3577#discussion_r27760820.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 2 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Ah right ... yeah, leave it as-is.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 4 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: I can revert.

I had an anonymous class at one point using this....but needed to get something from it so it is no longer anonymous.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 5 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: +1. Thanks!

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 6 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: +1

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java (hunk 8 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java]: So the thought is allow the iterator to abort early in a meaningful way if it needs/wants to.  We should be catching `Throwable` anyways.  Still remove?

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java]: Meh ... I'm fine with it as-is.

**Comment by Scottmitch** [comment]
@buchgr - Did you want to take a look...I think your PR https://github.com/netty/netty/pull/3516 has some iteration in it.  I can wait for you to finish that PR and absorb the changes if you want?

**Comment by buchgr** [comment]
@Scottmitch that would be great. thanks! I ll update my PR later today. I ll have a look at your PR tonight as well. Does that work for you?

**Comment by Scottmitch** [comment]
@buchgr works for me!

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 9]
```diff
@@ -391,4 +378,38 @@ void writeWindowUpdate(ChannelHandlerContext ctx) throws Http2Exception {
             ctx.flush();
         }
     }
+
+    /**
+     * Provides a means to iterator over all active streams and increment the flow control windows.
```
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: s/iterator/iterate

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 10]
```diff
@@ -226,10 +237,7 @@ private void writePendingBytes() {
             // Now write all of the allocated bytes. Copying the activeStreams array to avoid
```
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: @Scottmitch remove those comments.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 11]
```diff
@@ -23,16 +23,23 @@
...
 
 import java.util.ArrayDeque;
-import java.util.Collection;
 import java.util.Deque;
 
 /**
  * Basic implementation of {@link Http2RemoteFlowController}.
  */
 public class DefaultHttp2RemoteFlowController implements Http2RemoteFlowController {
+    private static final ActiveStreamVisitor WRITE_ON_ALL_ACTIVE_STREAMS = new ActiveStreamVisitor() {
```
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: @Scottmitch `WRITE_ALLOCATED_BYTES` ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 9 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: Done.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java (hunk 10 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: +1 :)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java (hunk 11 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: I like it. Done.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 12]
```diff
@@ -959,4 +1002,26 @@ private boolean isLocal() {
             return this == localEndpoint;
         }
     }
+
+    /**
+     * Allows events which would modify {@link DefaultHttp2Connection#activeStreams} to be queued while iterating over
+     * this collection.
+     */
+    private static final class ActiveStreamsEvent {
```
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @Scottmitch make interface

```java
    private interface ActiveStreamsEvent {
        void process();
    }
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 13]
```diff
@@ -376,14 +403,10 @@ public Http2Stream open(boolean halfClosed) throws Http2Exception {
...
-                createdBy().numActiveStreams++;
-
-                // Notify the listeners.
-                for (int i = 0; i < listeners.size(); i++) {
-                    listeners.get(i).onStreamActive(this);
-                }
+            if (canModifyActiveStreams()) {
+                addToActiveStreams();
+            } else {
+                activeStreamsPendingEvents.add(new ActiveStreamsEvent(this, false));
```
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @Scottmitch 

```java
                activeStreamsPendingEvents.add(new ActiveStreamsEvent() {
                    @Override
                    public void process() {
                        removeFromActiveStreams();
                    }
                });
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 14]
```diff
@@ -396,20 +419,13 @@ public Http2Stream close() {
...
-                        listeners.get(i).onStreamClosed(this);
-                    }
-                } finally {
-                    // Mark this stream for removal.
-                    removalPolicy.markForRemoval(this);
-                }
+            if (canModifyActiveStreams()) {
+                removeFromActiveStreams();
+            } else {
+                activeStreamsPendingEvents.add(new ActiveStreamsEvent(this, true));
```
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @Scottmitch 
```java
```java
activeStreamsPendingEvents.add(new ActiveStreamsEvent() {
    @Override
    public void process() {
        addToActiveStreams();
    }
});
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 12 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Done

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 15]
```diff
@@ -139,10 +139,17 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception { }
 
         public void channelInactive(ChannelHandlerContext ctx) throws Exception {
             try {
-                ChannelFuture future = ctx.newSucceededFuture();
-                final Collection<Http2Stream> streams = connection().activeStreams();
-                for (Http2Stream s : streams.toArray(new Http2Stream[streams.size()])) {
-                    closeStream(s, future);
+                final Http2Connection connection = connection();
+                // Check if there are streams to avoid the overhead of creating the ChannelFuture.
+                if (connection.numActiveStreams() != 0) {
```
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @Scottmitch why is this needed? what's `the overhead of creating the ChannelFuture.`?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 13 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Done.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 14 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Done.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java (hunk 8 above)
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java]: > So the thought is allow the iterator to abort early in a meaningful way if it needs/wants to.

can't it do that by just returning false?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 16]
```diff
@@ -63,9 +63,11 @@
     private final IntObjectMap<Http2Stream> streamMap = new IntObjectHashMap<Http2Stream>();
     private final ConnectionStream connectionStream = new ConnectionStream();
     private final Set<Http2Stream> activeStreams = new LinkedHashSet<Http2Stream>();
+    private final Queue<ActiveStreamsEvent> activeStreamsPendingEvents = new ArrayDeque<ActiveStreamsEvent>(4);
     private final DefaultEndpoint<Http2LocalFlowController> localEndpoint;
     private final DefaultEndpoint<Http2RemoteFlowController> remoteEndpoint;
     private final Http2StreamRemovalPolicy removalPolicy;
+    private int activeStreamsIterationCount;
```
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @Scottmitch 

I wonder if it makes sense to move all this active streams related stuff into its own class? Something like

```java
    private static class ActiveStreams {
        LinkedHashSet<Http2Stream> streams;
        Queue<ActiveStreamsEvent> pendingEvents = new ArrayDeque<ActiveStreamsEvent>();
        int pendingIterations;
        
        Http2Stream forEachActiveStream(ActiveStreamVisitor visitor) {
        ...
        }

        interface Event {
            void process();
        }
    }
```

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Seems reasonable.  Let me take a stab.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 15 above)
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @normanmaurer @Scottmitch hmm ... that seems like a super rare case to me. Can we change the condition to `> 0`?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 17]
```diff
@@ -391,4 +378,38 @@ void writeWindowUpdate(ChannelHandlerContext ctx) throws Http2Exception {
...
+        @Override
+        public boolean visit(Http2Stream stream) throws Http2Exception {
+            try {
+                // Increment flow control window first so state will be consistent if overflow is detected.
+                FlowState state = state(stream);
+                state.incrementFlowControlWindows(delta);
+                state.incrementInitialStreamWindow(delta);
+            } catch (StreamException e) {
+                if (compositeException == null) {
+                    compositeException = new CompositeStreamException(e.error(), DEFAULT_COMPOSITE_EXCEPTION_SIZE);
```
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: @Scottmitch I know it's been here before but do we really need the `DEFAULT_COMPOSITE_EXCEPTION_SIZE`?

**Comment by buchgr** [comment]
@Scottmitch great job! I like it.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 15 above)
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @buchgr  +1 ...

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 17 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: I guess not.  However it is nice to make it visible and provide a name along with the logical value.  It was more useful before when it was used in multiple spots...

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 15 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @buchgr - I will change but I have been told in other reviews to change `> 0` to `!= 0` for other "unsigned" types. (for example dealing with ByteBuf indices or readableBytes... @normanmaurer :) )

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java (hunk 8 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java]: > can't it do that by just returning false?

The false return is more designed to "find an interesting stream" not return a "meaningful" error that can be translated into a HTTP/2 type error.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 15 above)
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @Scottmitch uhm am I missing something ... why is that relevant for `numActiveStreams()`?

**Comment by Scottmitch** [comment]
@nmittler @buchgr - Care to take another look?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 18]
```diff
@@ -959,4 +940,139 @@ private boolean isLocal() {
             return this == localEndpoint;
         }
     }
+
+    /**
+     * Provides an interface to manage the active streams for this {@link Http2Connection}.
+     */
+    private interface ActiveStreams {
```
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @Scottmitch nit: we don't need the interface really? could we just have one `class ActiveStreams`?

**Comment by buchgr** [comment]
@Scottmitch just one comment. LGTM.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java (hunk 8 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java]: nit: Should we call this `StreamVisitor`? Maybe in the future we might want to also be able to iterate over all streams (not just active).

**Comment by nmittler** [comment]
@Scottmitch looks awesome! just one minor question ... otherwise LGTM. Feel free to cherry-pick.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 18 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: It allows the `Event` interface to nest inside of `ActiveStreams` where we couldn't put this nested `Event` interface inside the `DefaultActiveStreams`.  I can rename the `Event` interface to `ActiveStreamsEvent` and put it inside the `DefaultHttp2Connection` scope if that is preferred?

Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @Scottmitch why is that? Should be able to put the `Event` interface into `DefaultActiveStreams`?

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: `DefaultActiveStreams` is not static.

```bash
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] .../git/netty/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java:[977,8] error: Illegal static declaration in inner class DefaultHttp2Connection.DefaultActiveStreams
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java (hunk 8 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java]: @nmittler +1. Done.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 18 above)
Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @Scottmitch oh can't be static because of the listeners? Fine :-).

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Yah inner classes may be getting a bit out of hand in this class....

Comment (buchgr) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @Scottmitch maybe make the class static and pass in a reference to the connection?

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @buchgr - Was thinking about that, but we are using `this` in the constructor which is also awkward...not that we are necessarily avoiding the issue here but it is still not ideal....I'll make the change and commit to see how it looks.

**Comment by Scottmitch** [comment]
@buchgr @nmittler - WDYT about removing the `interface ActiveStream`?

**Comment by buchgr** [comment]
@Scottmitch you could also just pass in the listeners right?

**Comment by Scottmitch** [comment]
@buchgr - And the removal policy I guess...

**Comment by nmittler** [comment]
@Scottmitch still LGTM

**Comment by buchgr** [comment]
@Scottmitch awww fine :D LGTM

**Comment by Scottmitch** [comment]
Alright I'll cherry-pick after I pull in @buchgr's https://github.com/netty/netty/pull/3516 PR.

**Comment by Scottmitch** [comment]
Thanks for the review everyone!

Cherry-picked 4.1 (83ce8a9) master (28bd5a5).
